go/vt/vttablet/tabletmanager/vreplication/vcopier_copy_all.go

```diff
@@ -0,0 +1,297 @@
+func (vc *vcopier) copyAll(ctx context.Context, settings binlogplayer.VRSettings) error {
+	var err error
+
+	log.Infof("Starting copyAll for %s", settings.WorkflowName)
+	defer log.Infof("Returning from copyAll for %s", settings.WorkflowName)
+	defer vc.vr.dbClient.Rollback()
+
+	state, err := newCopyAllState(vc)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, copyPhaseDuration)
+	defer cancel()
+
+	rowsCopiedTicker := time.NewTicker(rowsCopiedUpdateInterval)
+	defer rowsCopiedTicker.Stop()
+	copyStateGCTicker := time.NewTicker(copyStateGCInterval)
+	defer copyStateGCTicker.Stop()
+
+	parallelism := getInsertParallelism()
+	copyWorkerFactory := vc.newCopyWorkerFactory(parallelism)
+	var copyWorkQueue *vcopierCopyWorkQueue
+
+	// Allocate a result channel to collect results from tasks.
+	resultCh := make(chan *vcopierCopyTaskResult, parallelism*4)
+	defer close(resultCh)
+
+	var lastpk *querypb.Row
+	var pkfields []*querypb.Field
+	var lastpkbv map[string]*querypb.BindVariable
+	// Use this for task sequencing.
+	var prevCh <-chan *vcopierCopyTaskResult
+	var gtid string
+
+	serr := vc.vr.sourceVStreamer.VStreamTables(ctx, func(resp *binlogdatapb.VStreamTablesResponse) error {
+		defer vc.vr.stats.PhaseTimings.Record("copy", time.Now())
+		defer vc.vr.stats.CopyLoopCount.Add(1)
+		log.Infof("VStreamTablesResponse: received table %s, #fields %d, #rows %d, gtid %s, lastpk %+v",
+			resp.TableName, len(resp.Fields), len(resp.Rows), resp.Gtid, resp.Lastpk)
+		tableName := resp.TableName
+		gtid = resp.Gtid
+
+		updateRowsCopied := func() error {
+			updateRowsQuery := binlogplayer.GenerateUpdateRowsCopied(vc.vr.id, vc.vr.stats.CopyRowCount.Get())
+			_, err := vc.vr.dbClient.Execute(updateRowsQuery)
+			return err
+		}
+
+		if err := updateRowsCopied(); err != nil {
+			return err
+		}
+		select {
+		case <-rowsCopiedTicker.C:
+			if err := updateRowsCopied(); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return io.EOF
+		default:
+		}
+		if tableName != state.currentTableName {
+			if copyWorkQueue != nil {
+				copyWorkQueue.close()
+			}
+			copyWorkQueue = vc.newCopyWorkQueue(parallelism, copyWorkerFactory)
+			if state.currentTableName != "" {
+				log.Infof("copy of table %s is done at lastpk %+v", state.currentTableName, lastpkbv)
+				if err := vc.deleteCopyState(state.currentTableName); err != nil {
+					return err
+				}
+			} else {
+				log.Infof("starting copy phase with table %s", tableName)
+			}
+
+			state.currentTableName = tableName
+		}
+		if !copyWorkQueue.isOpen {
+			if len(resp.Fields) == 0 {
+				return fmt.Errorf("expecting field event first, got: %v", resp)
+			}
+			fieldEvent := &binlogdatapb.FieldEvent{
+				TableName: tableName,
+			}
+			fieldEvent.Fields = append(fieldEvent.Fields, resp.Fields...)
+			tablePlan, err := state.plan.buildExecutionPlan(fieldEvent)
+			if err != nil {
+				return err
+			}
+			pkfields = append(pkfields, pkfields...)
```

Comment:
I don't understand the above line. Is it intentional? To double the slice? And this happens how many times?

Reply:
Oops. That is. not right. Fixed: need to get pkfields from the response. The reason it was not noticed  or caused any test failures is because it is only used in a log line.